lib/core/networking/websockets/middleware/encryption_middleware.dart

```diff
@@ -0,0 +1,216 @@
+import 'dart:async';
+import 'dart:convert';
+import 'dart:io';
+
+import 'package:cryptography/cryptography.dart';
+import 'package:synchronized/extension.dart';
+
+import 'middleware_types.dart';
+import '../session.dart';
+
+/// Middleware that handles secure communication
+Future<dynamic> encryptionMiddleware(
+  dynamic session,
+  dynamic message,
+  MiddlewareAction action,
+  Future<dynamic> Function(dynamic) next,
+) async {
+  final CryptoProvider provider = session.storage['crypto_provider'] ?? CryptoProvider(session);
+
+  // Redirect handshaking messages to the crypto provider
+  if (_isHandshakeMsg(message)) {
+    // Received handshaking messages are captured
+    if (action == MiddlewareAction.receive) {
+      provider(message);
+      return null;
+    }
+
+    // Handshaking messages are not blocked nor passed to other middleware
+    return message;
+  }
+
+  // Wait for secure channel to be established
+  await provider.established;
+
+  if (action == MiddlewareAction.receive) {
+    // We catch all errors here as we can't trust the incoming data
+    try {
+      // Decrypt the message so other middleware can work with it
+      message = await provider.decrypt(message);
+    } catch (e) {
+      session.raise(e);
+      return null;
+    }
+
+    return await next(message);
+  } else {
+    // Let other middleware work with the message, finally encrypt it
+    return await provider.encrypt(await next(message));
+  }
+}
+
+/// Check if message belongs to handshaking protocol
+bool _isHandshakeMsg(String msg) {
+  return msg.startsWith(_CryptoVars.prefix);
+}
+
+/// Class that handles secure communication
+/// between client and server.
+///
+/// Every session has it's own instance of this class.
+/// It normally lives in `session.storage['crypto_provider']`
+class CryptoProvider {
```

Comment:
For some reason my og. comment got lost I think. 

I get why you included `CryptoProvider` in this file. You are using it only here. Why public then? 
You also do not unit test the class itself. The only reason I can see why you might want this to be public is because you type check `storage['crypto-provider']`. But that's not enough to export the whole class. You could as good use `storage['crypto-provider'].runtimeType.toString()`

Reply:
I’ll look into this. We may need the import for testing.